Small features (#415)
ingonyama-zk/icicle: docs/docs/icicle/supporting-additional-curves.md

### Adding curve_name_params.cuh

Start by copying `bn254_params.cuh` contents in your params file. Params should include:
 - **fq_config** - parameters of the Base field.
    - **limbs_count** - `ceil(field_byte_size / 4)`.
    - **modulus_bit_count** - bit-size of the modulus.
    - **num_of_reductions** - the number of times to reduce in reduce function.
    - **modulus** - modulus of the field.
    - **modulus_2** - modulus * 2.
    - **modulus_4** - modulus * 4. 
    - **neg_modulus** - negated modulus. 
    - **modulus_wide** - modulus represented as a double-sized integer.
    - **modulus_squared** - modulus**2 represented as a double-sized integer.
    - **modulus_squared_2** - 2 * modulus**2 represented as a double-sized integer.
    - **modulus_squared_4** - 4 * modulus**2 represented as a double-sized integer.
    - **m** - value used in multiplication. Can be computed as `2**(2*modulus_bit_count) // modulus`. 
    - **one** - multiplicative identity. 
    - **zero** - additive identity. 

>     - **montgomery_r** - `2 ** 256 % modulus`

Author: good catch, I missed that

    - **montgomery_r_inv** - `2 ** (-256) % modulus`
 - **fp_config** - parameters of the Scalar field.
    Same as fq_config, but with additional arguments:
    - **omegas_count** - [two-adicity](https://cryptologie.net/article/559/whats-two-adicity/) of the field. And thus the maximum size of NTT.
    - **omegas** - an array of omegas for NTTs. An array of size `omegas_count`. The ith element is equal to `1.nth_root(2**(2**(omegas_count-i)))`.
    - **inv** - an array of inverses of powers of two in a field. Ith element is equal to `(2 ** (i+1)) ** -1`.
 - **G1 generators points** - affine coordinates of the generator point.
 - **G2 generators points** - affine coordinates of the extension generator. Remove these if `G2` is not supported.
 - **Weierstrass b value** - base field element equal to value of `b` in the curve equation.
 - **Weierstrass b value G2** - base field element equal to value of `b` for the extension. Remove this if `G2` is not supported.
 

> Note: All the params are not in Montgomery form. To convert number values into `storage` type you can use this python function:

Reviewer: Lets change this to an admonition/callout: https://docusaurus.io/docs/markdown-features/admonitions

> To add support a new curve you must create a new file under [`icicle/curves`](https://github.com/ingonyama-zk/icicle/tree/main/icicle/curves). The file should be named `<curve_name>_params.cuh`.

Reviewer 2: ```suggestion
To add support for a new curve you must create a new file under [`icicle/curves`](https://github.com/ingonyama-zk/icicle/tree/main/icicle/curves). The file should be named `<curve_name>_params.cuh`.
```

### curve_name_params.cuh

Start by copying `bn254_params.cuh` contents in your params file. Params should include:
 - **fq_config** - parameters of Base field.
    - **limbs_count** - `ceil(field_byte_size / 4)`.
    - **modulus_bit_count** - bit-size of the modulus.
    - **num_of_reductions** - the number of times to reduce in reduce function.
    - **modulus** - modulus of the field.
    - **modulus_2** - modulus * 2.
    - **modulus_4** - modulus * 4. 
    - **neg_modulus** - negated modulus. 
    - **modulus_wide** - modulus represented as a double-sized integer.
    - **modulus_squared** - modulus**2 represented as a double-sized integer.
    - **modulus_squared_2** - 2 * modulus**2 represented as a double-sized integer.
    - **modulus_squared_4** - 4 * modulus**2 represented as a double-sized integer.
    - **m** - value used in multiplication. Can be computed as `2**(2*modulus_bit_count) // modulus`. 
    - **one** - multiplicative identity. 
    - **zero** - additive identity. 
    - **montgomery_r** - `2 ** 256 % modulus`
    - **montgomery_r_inv** - `2 ** (-256) % modulus`

>  - **fp_config** - parameters of Scalar field.

Reviewer 2: parameters of the scalar field

add the `the`

set(SUPPORTED_CURVES bn254;bls12_381;bls12_377;bw6_761;<curve_name>)
```

### Poseidon

If you want your curve to implement Poseidon hash function or a tree builder, you will need to pre-calculate the optimized parameters for it.  

> Copy [constants_template.h](https://github.com/ingonyama-zk/icicle/blob/main/icicle/appUtils/poseidon/constants/constants_template.h) into `icicle/appUtils/poseidon/constants/<CURVE>_poseidon.h`. Run the [constants generation script](https://dev.ingonyama.com/icicle/primitives/poseidon#constants). The script will print the number of partial rounds and generate `constants.bin` file. Use `xxd -i constants.bin` to parse the file into C declarations. Copy the `unsigned char constants_bin[]` contents inside your new file. Repeat this process for arities 2, 4, 8 and 11.

Reviewer 2: ```suggestion
Copy [constants_template.h](https://github.com/ingonyama-zk/icicle/blob/main/icicle/appUtils/poseidon/constants/constants_template.h) into `icicle/appUtils/poseidon/constants/<CURVE>_poseidon.h`. Run the [constants generation script](https://dev.ingonyama.com/icicle/primitives/poseidon#constants). The script will print the number of partial rounds and generate a `constants.bin` file. Use `xxd -i constants.bin` to parse the file into C declarations. Copy the `unsigned char constants_bin[]` contents inside your new file. Repeat this process for arities 2, 4, 8 and 11.
```

After you've generated the constants, add your curve in this [SUPPORTED_CURVES_WITH_POSEIDON](https://github.com/ingonyama-zk/icicle/blob/main/icicle/CMakeLists.txt#L72) in the `CMakeLists.txt`.

## Bindings

> In order to support a new curves in the binding libraries you first must support it in ICICLE core.

Reviewer: ```suggestion
In order to support a new curve in the binding libraries you first must support it in ICICLE core.
```

### Rust

> Go to [rust wrappers folder](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/rust/icicle-curves) and copy `icicle-curve-template` to a new folder named `icicle-<curve_name>`.

Reviewer: I think it is confusing to call `wrappers/rust/icicle-curves` the rust wrappers folder. I would assume the `wrappers/rust` would be called that.

Find all the occurrences of `<CURVE>` placeholder inside the crate. (You can use `Ctrl+Shift+F` in VS Code or `grep -nr "<CURVE>"` in bash). You will then need to replace each occurrence with your new curve name.

#### Limbs

> Go to [curve file](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/rust/icicle-curves/icicle-curve-template/src/curve.rs) and set `SCALAR_LIMBS`, `BASE_LIMBS` and `G2_BASE_LIMBS` (if G2 is needed) to a minimum number of `u64` required to store a single scalar field / base field element respectively.  

Reviewer: This will likely cause issues where someone will change the limbs in the template file and not their newly created curve.rs file.

```suggestion
Go to your curve's `curve.rs` file and set `SCALAR_LIMBS`, `BASE_LIMBS` and `G2_BASE_LIMBS` (if G2 is needed) to a minimum number of `u64` required to store a single scalar field / base field element respectively.  
```
